app/controllers/automated_tests_controller.rb

```diff
@@ -35,14 +35,26 @@ def student_interface
     @student = current_role
     @grouping = @student.accepted_grouping_for(@assignment.id)
 
-    unless @grouping.nil?
+    if @grouping.present?
       @grouping.refresh_test_tokens
       @authorized = flash_allowance(:notice,
                                     allowance_to(:run_tests?,
                                                  current_role,
                                                  context: { assignment: @assignment, grouping: @grouping })).value
-    end
 
+      if @assignment.enable_student_tests
+        if @assignment.non_regenerating_tokens
```

Comment:
It's great that you've simplified the logic; now, you can simplify it even further by combining this condition with the one above (i.e., `if @assignment.enable_student_tests && !@assignment.non_regenerating_tokens`). You don't need a separate branch to assign `@next_token_generation_time = nil` here because it'll be set in the "catch-all" statement on Line 57.